prov/tcp/src/tcpx_progress.c

```diff
@@ -209,19 +215,178 @@ static void progress_tx_queue(struct tcpx_ep *ep)
 	fastlock_release(&ep->queue_lock);
 }
 
-void tcpx_manual_progress(struct util_ep *util_ep)
+static int tcpx_auto_prog_rx_queues(struct tcpx_prog_mgr *mgr)
 {
 	struct tcpx_ep *ep;
+	int ret, i;
+
+	fastlock_acquire(&mgr->lock);
+	ret = fi_epoll_wait(mgr->epoll_fd, mgr->ctxs,
+			    MIN(mgr->used, mgr->ctxs_sz), 0);
+
+	if (ret < 0) {
+		FI_WARN(&tcpx_prov, FI_LOG_DOMAIN,
+			"fi_epoll_wait failed\n");
+		fastlock_release(&mgr->lock);
+		return -errno;
+	}
 
-	ep = container_of(util_ep, struct tcpx_ep, util_ep);
+	for (i = 0; i < ret ; i++) {
+		ep = (struct tcpx_ep *) mgr->ctxs[i];
+		progress_rx_queue(ep);
+	}
+	fastlock_release(&mgr->lock);
+	return FI_SUCCESS;
+}
 
-	progress_tx_queue(ep);
-	progress_rx_queue(ep);
+static void tcpx_auto_prog_tx_queues(struct tcpx_prog_mgr *mgr)
+{
+	struct dlist_entry *entry;
+	struct tcpx_ep *ep;
+	struct fid_list_entry *item;
+
+	fastlock_acquire(&mgr->lock);
+	dlist_foreach(&mgr->ep_list,entry) {
+		item = container_of(entry, struct fid_list_entry, entry);
+		ep = container_of(item->fid, struct tcpx_ep,
+				  util_ep.ep_fid.fid);
+		progress_tx_queue(ep);
+	}
+	fastlock_release(&mgr->lock);
+}
+
+void tcpx_manual_progress(struct util_ep *util_ep)
+{
+	struct tcpx_domain *domain;
+
+	domain = container_of(util_ep->domain,
+			      struct tcpx_domain,
+			      util_domain);
+
+	tcpx_auto_prog_tx_queues(&domain->progress.mgr);
+	tcpx_auto_prog_rx_queues(&domain->progress.mgr);
```

Comment:
Manual progress should really be associated with a CQ.  The utility code has callbacks for this, including linking into wait sets and poll sets.

Reply:
I will make the changes aptly.